### ISSUE TITLE
Input sources order and complex layouts

### ISSUE THREAD

**Issue by laughedelic** [noraesae/kawa#2]
Hi! It's a nice program. But I have a couple of problems. I have three layouts and set hotkeys as follows: 
- U.S. Extended `ctrl-1`
- Russian `ctrl-2`
- Pinyin - Simplified  `ctrl-3`

I get a wrong order of input sources, shifted in cycle by one: english-2, russian-3, chinese-1.

The second problem is with the pinyin input source, if I switch to it by the kawa hotkey, it doesn't work properly: the suggestions don't pop out until I switch to it using system hotkey. I've turned on the advanced method setting, but it doesn't change anything.

**Comment by noraesae** [comment]
Yes, I've recently encountered the same issue. I'll try to address the issue as soon as possible, but it's kinda tricky so it may take some time.

Sorry and thank you for the report.

**Comment by laughedelic** [comment]
hi @noraesae. is there any update on this issue?

**Comment by noraesae** [comment]
Hi, I'm afraid no. Currently I don't have enough time to contribute on Kawa. Fortunately, this is happening in my machine too, so it's at least much easier to fix now, but maybe take some more time. Sorry!

**Comment by laughedelic** [comment]
Ok, I understand. But could you at least make a release with the fix from #4?

**Comment by noraesae** [comment]
Thinking about trying to fix this issue today. If I can, I will release it with #4 this night(UTC+9). If I cannot, will release just #4.

**Comment by laughedelic** [comment]
Cool! Good luck 👍

**Comment by noraesae** [comment]
I've worked on this and tried several methods. It's all about come up with a workaround for the macOS bug concerning `TIS` library.

TL;DR: I decided to remove 'advanced input switching' and just wait the fix from Apple.

## Actual Problem

The actual problem is that switching by `TIS` APIs doesn't work well with CJS or complex input sources. The flag icon on the top right corner is changed, but the key input itself becomes different from the layout the flag indicates.

## Possible workarounds (and failures)

After investigating, I found that changing layout with OS native shortcuts(e.g. `command-shift`) seems like fixing the problem. So I tried the followings.

- Choose a layout one step ahead from the goal layout, and use `Select next source in Input menu` shortcut: The previous implementation of 'advanced switching'. It failed because the input source list returned by `TISCreateInputSourceList` is completely unordered, different from the actual order in the `Keyboard` preference, and it results in switching to a wrong layout.
- Repeat `Select next source in Input menu` until the goal language is current: The current input return by `TISCopyCurrentKeyboardInputSource` is *sometimes* not changing with no reason at all. It doesn't matter to try checking it after a specific timeout or even the `kTISNotifySelectedKeyboardInputSourceChanged` notification.
- Simulating `Select the previous input source` twice: It doesn't work either. Not switching as expected.

## Deprecation

All in all, `TIS` seems like being deprecated(actually `Carbon` itself is being deprecated). There is no doc provided by Apple. I expect Apple provide a modern API regarding input source soon, with full support of Swift.

## Conclusion

As I said in TL;DR, I may not try solving this as it seems extremely tricky, and just wait for the new API or at least the fix coming out from Apple.

P.S. #4 is published through v0.1.3.

**Comment by laughedelic** [comment]
@noraesae thanks for explanation and the published release.

**Comment by zcbenz** [comment]
@noraesae I have been troubled with the bug of TIS when trying to write a input method switcher, and then I found your awesome project!

So on this bug, I think there is one workaround that can actually work. The input method list is ordered by most recently used, so if we first switch to the target language, and then switch to another language, and then simulate "Select the previous input source", it should be guaranteed to work.

The logic is like this:
* If target is English
  * Select English
* If target is a language other than English
  1. Switch to target language
  2. Switch to English
  3. Select previous language

This trick has been used by
https://github.com/CodeFalling/fcitx-remote-for-osx/blob/7f7e2895d413df44f45ef5c9aa3d127e428a7cbc/fcitx-remote/main.m#L69
and https://github.com/tekezo/Karabiner/issues/308#issuecomment-190693550 so I think it should work.

I would like to write a pull request but I do not know Swift.

**Comment by noraesae** [comment]
@zcbenz Thanks for your comment. I'll create a separate issue and see if I can make it work.
